File: Contest/MaximumBagsWithFullCapacityofRocks.java
class Solution {

	// TC : O(n)
	// SC : O(n)
	public int maximumBags(int[] capacity, int[] rocks, int additionalRocks) {
		int[] vacancy = new int[capacity.length];
		for(int i=0;i<capacity.length;i++) {
			vacancy[i] = capacity[i] - rocks[i];
		}


		Arrays.sort(vacancy);
		int count = 0;
		for(int i=0;i<capacity.length;i++) {
			if(vacancy[i] == 0){
				count++;
			} else {
				if(additionalRocks>=vacancy[i]) {
					additionalRocks -= vacancy[i];
					count++;
				} else {
					break;
				}
			}
		}
		return count;
	}
}

File: December2022/C++/MaximumBagsWithFullCapacityOfRocks.cpp
// https://leetcode.com/problems/maximum-bags-with-full-capacity-of-rocks/
class Solution {
public:
    // TC: O(nlogn)
    // SC: O(n)
    int maximumBags(vector<int>& capacity, vector<int>& rocks, int additionalRocks) {
        int n = capacity.size();
        vector<int> remainingCapacity(n);

        for (int i = 0; i < n; i++) {
            remainingCapacity[i] = capacity[i] - rocks[i];
        }
        sort(remainingCapacity.begin(), remainingCapacity.end());

        int ans = 0;
        for (int i = 0; i < n; i++) {
            if (additionalRocks >= remainingCapacity[i]) {
                additionalRocks -= remainingCapacity[i];
                ans++;
            } else {
                break;
            }
        }

        return ans;
    }
};
